principles.rst

```diff
@@ -303,3 +309,130 @@ down GHC even when those features are not active. Yet this principle is importan
 as we hope not to make the current situation worse.
 
 From `#378`_, slightly generalized.
+
+
+
+GHC stability principles
+--------------------------
+
+The Haskell ecosystem has a built-in tension between stability and innovation.
+
+One the one hand, breaking changes impose heavy costs:
+
+* Users keep having to update their code
+* Library authors keep having to update their code
+* Updating to a later version of a library can in turn force taking on new dependencies.
+
+These difficulties add friction, discourage adoption, and make Haskell an unpredictable investment.
+
+On the other hand, GHC has always sought to be a laboratory for innovation.  If the stability guarantees are too onerous, we risk imposing a different set of costs:
+
+* Volunteers may get discouraged by the hoops they have to jump through to get a change agreed.
+* The language may become stuck in a local optimum, because moving to a better design would require breaking changes.
+* GHC may become encrusted with cruft and workarounds that exist to preserve existing, but undesirable, behaviour.
+
+We can't navigate this tension with simple all-or-nothing rules.  We have to take each case on its merits, aware of both sets of costs.  Rather than starting the debate from scratch each time, we can have some general rules, or principles, to guide us.  The `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_ made a start in this direction; this section discusses GHC itself.
+
+Assumptions
+~~~~~~~~~~~~
+
+We assume:
+
+* That we have adopted the  `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_, which establishes the ``ghc-experimental`` package.
+* That we have identified some extensions as "Experimental".  These extensions may well change, and are subject to
+  much weaker stability goals.  `GHC Proposal #601 <https://github.com/ghc-proposals/ghc-proposals/pull/601>`_, describes a plan along these lines, but all we need for this document is a list of Experimental extensions.
+
+Terminology
+~~~~~~~~~~~~~~~
+
+We define a **stable Haskell package** as follows. A stable Haskell package
+
+* Does not use Experimental extensions.
+* Does not use experimental features.  (Examples: use of the Javascript or Wasm back end, builds on non-tier-1 platforms.)
+* Does not rely on explicitly-undefined behaviour. (Example: ``INCOHERENT`` instance selection.)
+* Does not use ``-Werror`` in its default build configuration.
+* Explicitly specifies a language edition (``Haskell98``, ``GHC2021``), in the source code or the build configuration.
+
+Stability (GR1)
+~~~~~~~~~~~~~~~~
+
+**General rule (GR1)**.  *A stable Haskell package P that works with GHC X should continue to work in subsequent releases of GHC, say GHC Y, provided P's dependencies work with GHC Y with only minor version bumps.*
```

Comment:
"Latest releases" of dependencies seems too restrictive to me, given that P may be on an older major version of a dependency and the latest major release of that dependency may itself have breaking changes. I think that's what "minor version bumps" was trying to get at, but I agree it could be clarified.

Perhaps something like "provided that for each of P's dependencies, the latest release of the dependency that is compatible with P's version bounds works with GHC Y". Of course this implicitly assumes P has accurate version bounds (i.e. follows PVP) but I guess that belongs in the definition of "stable Haskell package".

Reply:
It still does not make sense to me, the property you are talking about seems to be a property of a build plan, not of a package. A package itself does not have a single prescribed build plan, so "minor version bumps" is meaningless. 

I am aware of the issue of ecosystem upgrading in lockstep, but I'd suggest not to tackle it in this document.

Reply:
> It still does not make sense to me, the property you are talking about seems to be a property of a build plan, not of a package. A package itself does not have a single prescribed build plan, so "minor version bumps" is meaningless.

I see the difficulty but can you suggest a better way to phrase (GR1).  The intent is:
* If P's direct dependencies work with GHC Y, then so does P.

But that's a bit too vague.  Consider one of P's direct dependencies, Q.  It would be no good if the GHC-Y compatible version of Q exposes many more functions (major version bump), because that might break P.   How can we say that in a precise way?

Maybe it's best not to be too precise and stick with "If P's direct dependencies work with GHC Y, then so does P", but add some qualifying text to clarify?

Anyone?

Reply:
>  It would be no good if the GHC-Y compatible version of Q exposes many more functions (major version bump), because that might break P. 

* True, but that's not really for GHC SC to tackle.
* According to PVP exposing "many more functions" is not necessarily a major version bump.

I'd go for "If P can be compiled against the latest versions of its direct dependencies with GHC X (!), and all latest versions of its direct dependencies work with GHC Y, then so does P". This is a less challenging notion than you and Adam seem to pursue, but precise and actionable.

Reply:
Hmm.  Suppose P worked perfectly fine with dependency Q-2.3.1.   Then the author of Q releases Q-2.4, and P breaks.  Well, the author of P might well put an upper bound of Q-2.3.1. 

Now a new release of GHC comes out, GHC-Y.  It would be desirable if P still worked when compiled with GHC-Y, would it not?.  But your rule offers no such guarantee, because it is not true that "P can be compiled against the latest version of Q". 

We cannot promise that Q-2.3.2 will be compatible with GHC-Y.  But if we could make it compatible with only a minor bump, to Q-2.3.2 say, then one might hope it would work.

Still, the weaker the guarantee, the easier it is to fulfill!  And it is more complicated to explain the stronger guarantee.

Reply:
I believe that the weaker guarantee brings you 90% of potential benefit if fullfilled. You can always refine it later if you find the weaker notion too easy to achieve ;)

> We cannot promise that Q-2.3.2 will be compatible with GHC-Y. But if we could make it compatible with only a minor bump, to Q-2.3.2 say, then one might hope it would work.

Practically if there is `Q-2.4` already, `Q-2.3.2` almost certainly won't happen. Very few libraries care to maintain older major release series, there is zero incentive to do so unless you are paid. So I would not bother with a stronger guarantee, at least until our ecosystem is managed differently.

Reply:
> Then it is possible to release Q-2.3.1.1 with the same API as Q-2.3.1 but compatible with GHC Y. 

As I mentioned above, very few packages would bother to release Q-2.3.X after Q-2.4. I know zero examples of it outside of few boot libraries (who begrudginly must do it sometimes). 

The advantage is an easy-to-understand and easy-to-apply definition, expressed in terms of packages. I don't think the more elaborate definition (which is yet to be written) buys you anything practically valuable, but I have no skin in the game; if GHC SC feels well equipped to discuss PVP intricacies, you do you.

Reply:
I think it would be reasonable to say something weaker still that doesn't make any reference to the "latest" versions of anything:

```
A stable Haskell package P built against a set of package-version dependencies D that works with GHC X should continue to work in subsequent releases of GHC, say GHC Y, provided that the package-versions in D work with GHC Y.
```

It's not GHC's job to make sure that P's dependencies have released versions that are compatible with GHC Y. Perhaps the project is dead, perhaps the latest version is a total rewrite and they don't want to make a minor release. I think it's enough that GHC promises that _it's_ bit of the job, which is compiling your package given a set of pre-built dependencies, behaves the same. 

So the overall process might look like this:

1. P builds with GHC X, and depends on D-1, which works with GHC X but not GHC Y
2. P updates to depend on D-2, which works with both GHC X and Y
3. P builds with GHC Y

GR1 says that 2 and 3 should behave the same. Whether 1 and 2 behave the same (and how to fix it if they don't) is a problem for the maintainers of D and P.

Of course, if GR1 applies pointwise throughout the dependency tree, we might well hope that many of P's dependencies won't need version bumps to support GHC Y, and so we have fewer 1-2 steps to worry about. But I don't think we should make avoiding them part of GR1.